packages/artifact/src/internal/upload/zip.ts

```diff
@@ -30,31 +31,57 @@ export async function createZipUploadStream(
     `Creating Artifact archive with compressionLevel: ${compressionLevel}`
   )
 
-  const zip = archiver.create('zip', {
-    highWaterMark: getUploadChunkSize(),
-    zlib: {level: compressionLevel}
-  })
+  const zlibOptions = {
+    zlib: {
+      level: compressionLevel,
+      bufferSize: getUploadChunkSize()
+    }
+  }
+  const zip = new ZipStream.default(zlibOptions)
 
+  const bufferSize = getUploadChunkSize()
+  const zipUploadStream = new ZipUploadStream(bufferSize)
+  zip.pipe(zipUploadStream)
   // register callbacks for various events during the zip lifecycle
   zip.on('error', zipErrorCallback)
   zip.on('warning', zipWarningCallback)
   zip.on('finish', zipFinishCallback)
   zip.on('end', zipEndCallback)
-
-  for (const file of uploadSpecification) {
+  const addFileToZip = (
+    file: UploadZipSpecification,
+    callback: (error?: Error) => void
+  ): void => {
     if (file.sourcePath !== null) {
-      // Add a normal file to the zip
-      zip.append(createReadStream(file.sourcePath), {
-        name: file.destinationPath
-      })
+      zip.entry(
+        createReadStream(file.sourcePath),
+        {name: file.destinationPath},
+        (error: unknown) => {
+          if (error) {
+            callback(error as Error) // Cast the error object to the Error type
+            return
+          }
+          callback()
+        }
+      )
     } else {
-      // Add a directory to the zip
-      zip.append('', {name: file.destinationPath})
+      zip.entry('', {name: file.destinationPath}, (error: unknown) => {
+        if (error) {
+          callback(error as Error)
+          return
+        }
+        callback()
+      })
     }
   }
 
-  const bufferSize = getUploadChunkSize()
-  const zipUploadStream = new ZipUploadStream(bufferSize)
+  async.eachSeries(uploadSpecification, addFileToZip, (error: unknown) => {
+    if (error) {
+      core.error('Failed to add a file to the zip:')
+      core.info(error.toString()) // Convert error to string
+      return
+    }
+    zip.finalize() // Finalize the archive once all files have been added
+  })
```

Comment:
instead of introducing a new dependency on `async`, can we just await on a group of these promises?

e.g. 

```js
await Promise.all(uploadSpecification.map(addFileToZip))
```

or something like that? `eachSeries` [does one at a time](https://github.com/caolan/async/blob/92ea2182de27d3e6a02d79562ba562e3c6d59b88/lib/eachSeries.js#L5), does it have to do that for a specific reason? and do we know if this will impact speed?